[WIP] 6TiSCH Operation Sublayer (6top)

yatch opened the pull request:
# Summary
This PR aims to add 6TiSCH Operation Sublayer (6top) on Contiki.

[The original work](https://github.com/contiki-os/contiki/pull/1898) was done by Shalu Rajendran and Lijo Thomas. Some enhancements have been made, including supporting [draft-03 of 6P](https://tools.ietf.org/html/draft-ietf-6tisch-6top-protocol-03).

The status of code is work-in-progress. Please note that the code structures or APIs may be changed. As a matter of fact, `sixtop_sf_t` is going to have new members, `sixtop.h` will have new wrapper APIs for TSCH slotframe and link operations.

At this moment, there is no Schedule Function (SF) shipped with these changes. You may have to implement your own SF for dynamic cell allocation. Currently, I'm implementing [SF0](https://tools.ietf.org/html/draft-ietf-6tisch-6top-sf0-02) on this code base.

While there are many commits remaining on the branch, they would be squashed at some point.

# How to Play
An example code is provided under `example/ipv6/rpl-tsch-sixtop`:
- `rpl-tsch-sixtop-z1.csc`: a simulation file; a good place to start
- `sf-simple.[ch]`: a sample/simple SF implementation (no cell/bandwitdh estimation)
- `node-sixtop.c`: a node process using sf-simple to allocate/deallocate cells

# References
- The original PR by @shalurajendran : https://github.com/contiki-os/contiki/pull/1898
- IETF 6TiSCH Working Group: https://datatracker.ietf.org/wg/6tisch
- 6top Protocol (6P): https://tools.ietf.org/html/draft-ietf-6tisch-6top-protocol
- 6top SF0: https://tools.ietf.org/html/draft-ietf-6tisch-6top-sf0

simonduq commented on core/net/mac/tsch/sixtop/README.md:
```diff
@@ -0,0 +1,94 @@
+## Caveat
+
+At this moment, allocated 6P neighbor states never be deleted since 6P has no
+way to decide to delete them. 6P neighbor state consists of SeqNum to be used
```
how about the DELETE command??

yatch replied:
The Delete command doesn't help for this purpose. Even if a node has no cells with its peer as a result of the Delete command, it has to keep the neighbor state for the peer in order to:
- have a proper SeqNum value for the next request, whose value is obtained by **incrementing the previous SeqNum by exactly one**.
- keep GTX and GRX to avoid schedule generation inconsistency in the following transactions

To make matters worse, the node would keep the neighbor state even if the peer is gone...

simonduq replied:
Right, I see. I misread and thought this was about cells, not nbr state.

simonduq commented on core/net/mac/tsch/sixtop/sixp.c:
```diff
@@ -0,0 +1,967 @@
+
+static sixp_nbr_t *find_nbr(const linkaddr_t *addr);
+static sixp_nbr_t *alloc_nbr(const linkaddr_t *addr);
+#if 0
```
please remove dead code or add config parameter instead

yatch replied:
Yep! I guess we will not need them once `nbr-table` is applied here.

simonduq commented on core/net/mac/tsch/sixtop/sixp.c:
```diff
@@ -0,0 +1,967 @@
+MEMB(trans_memb, sixp_trans_t, SIXTOP_6P_MAX_TRANSACTIONS);
+LIST(trans_list);
+
+MEMB(nbr_memb, sixp_nbr_t, SIXTOP_6P_MAX_NEIGHBORS);
```
I think so too, but please note that nbr-table is to be used only outside of interrupt (read and write). I think that should be fine with 6p but just checking.

simonduq commented on core/net/mac/tsch/sixtop/sixp.c:
```diff
@@ -0,0 +1,967 @@
+}
+/*---------------------------------------------------------------------------*/
+static sixp_nbr_t *
+find_nbr(const linkaddr_t *addr)
```
ok that will be replaced with the nbr table lookup then!

simonduq commented on core/net/mac/tsch/sixtop/sixp.c:
```diff
@@ -0,0 +1,967 @@
+  return nbr;
+}
+/*---------------------------------------------------------------------------*/
+#if 0
```
dead code here again

simonduq commented on core/net/mac/tsch/sixtop/sixp.h:
```diff
@@ -0,0 +1,297 @@
+  uint8_t cell_options;  /**< CellOptions */
+  uint8_t num_cells;     /**< NumCells */
+  uint32_t cell_list[0]; /**< CellList */
+} __attribute__((packed)) sixp_req_add_t;
```
Pretty sure this had portability issues (across platforms and/or compilers).
The practice in uip6 is not to use struct for packet formatting. Fields are read/written one by one at a position guaranteed to be correct. Less efficient but 100% portable. For consistency with the rest of contiki I think 6p should follow the same practice.

simonduq commented on core/net/mac/tsch/sixtop/sixp.h:
```diff
@@ -0,0 +1,297 @@
+} sixp_send_status_t;
+
+/**
+ * \brief 6P Add Request Format
```
shouldn't format go to `sixp-packet`?

simonduq commented on core/net/mac/tsch/sixtop/sixp-packet.c:
```diff
@@ -0,0 +1,248 @@
+  packetbuf_clear();
+
+  /* put 6top SubIE Content into packetbuf */
+  if(create_6top_ie(ie) < 0){
```
missing space before `{`

simonduq commented on core/net/mac/tsch/sixtop/sixp.c:
```diff
@@ -0,0 +1,967 @@
+   * Section 4.3.3, draft-ietf-6tisch-6top-protocol-03.
+   *
+   * The assumption here is that there is one transactions for a single peer at
+   * most.
```
But doesn't the I-D also state that concurrent 6P transactions with a given neighbor are not possible?
Here, we have multiple transactions, but only one per nbr, right?

yatch replied:
Sorry, I overlooked this comment somehow... [Section 4.3.3](https://tools.ietf.org/html/draft-ietf-6tisch-6top-protocol-03#section-4.3.3) says, "only a single 6P Transaction between two neighbors, in a given direction, can take place at the same time." Therefore, a node may have at most two transactions with the same peer, one of which is initiated by the node, the other is by the peer. The same thing is said by "nodes A and B MAY support having two transactions going on at the same time, one in each direction."

yatch commented:
@simonduq Thank you for the review. I guess I'll have time to work on this PR the latter half of this week..!

yatch commented on core/net/mac/tsch/sixtop/sixtop.c:
```diff
@@ -0,0 +1,204 @@
+  assert(llsec_input != NULL);
+
+  /*
+   * A received *DATA* frame is supposed to be stored in packetbuf by
```
Thanks, this is what I forgot to ask of the 6TiSCH WG.

The current implementation could handle a 6P message, for example, in an EACK frame. But, I think it's not intended by the 6P draft. Any type of frame can have Payload IE. So, I think we need to define which type of frame can have an IETF IE.

yatch commented on core/net/mac/tsch/sixtop/sixtop.c:
```diff
@@ -0,0 +1,204 @@
+  }
+
+  /* Pass to the upper layer */
+  llsec_input();
```
I think something for upper layers could be in Frame Payload.

yatch replied:
This is because it's the upper layer, next layer, of `6top` in this case.

It's just a difference of ways of thinking. If we consider 6top as a part of the TSCH layer, the upper layer (`llsec_input()`) is called at the caller, TSCH layer. In the current implementation, 6top is considered as a separate layer from TSCH. So, it calls the upper layer by itself.

Both could work.

yatch commented:
@simonduq Just pushed the snapshot of the branch including updates to address your comments. Now it uses `nbr-table` to maintain `sixp_nbr`. In addition, it does not use any structure to parse or create a 6P packet.

Thank you for your review. I'll test this code with implementing SF0.

simonduq commented:
Please let me know when this is ready for a final review.

yatch commented:
@simonduq Sure!

sieben commented:
Me too :-) I want to perform 6P interop testing between Contiki and OpenWSN.

yatch commented:
@siben I'll write here something to share, although I've been too busy to write 6P code lately...

simonduq commented:
My recommendation is not to set the bar high when it comes to more features: stick to what you have now, when you're happy with the code quality I'll review and once this is merged more features can be added later. Else the PR might become huge and take many months to get merged..

yatch commented:
@simonduq Thank you for the comment! I got it.

yatch commented:
I'm sorry that it takes so long... I think I can update the branch next week. Basically, this update doesn't have any functional additions but only refactoring things. Stay tuned... :radio:

yatch commented:
I think it's ready for code review. Major changes since the last push, January are:

- API clean-ups
- Change file structure; introducing `sixp-pkt.[ch]`, `sixp-nbr.[ch]`, `sixp-trans.[ch]`, and `sixtop-conf.h`
- Introduce regression tests for the 6top and 6P implementation (`regression-tests/26-6tisch`)
- Bugfix; append Payload Termination IE after 6top IE Content (follows https://tools.ietf.org/html/draft-ietf-6tisch-6top-protocol-03#section-4.2.1)
- Add `nbr_table_is_register()`. This change, https://github.com/contiki-os/contiki/pull/1981/commits/a78fbe843b9f8d53b018a348fac6fa198e1433f3, would be a separate PR.

`sf-simple` is implemented under `examples/ipv6/rpl-tsch-sixtop/` and works as before.

simonduq commented:
Wonderful! Am having busy days at the moment but will find the time to review eventually. BTW there seems to be a conflict with master.

yatch commented:
@simonduq Thanks. On my side, super busy days have gone 😄 

With regard to the conflict, I'm aware of it and, of course, it needs to be resolved. I may push some minor changes in near future. In such a case, I won't do force-push so that we can see what changes are made after "ready for review".

I'm working on draft-04 support right now.

sumanpanchal commented on core/net/mac/tsch/sixtop/README.md:
```diff
@@ -0,0 +1,103 @@
+
+## Example
+
+An example using 6top is provided under `examples/ipv6/rpl-tsch-sxitop/`.
```
Typo error: `examples/ipv6/rpl-tsch-sxitop/`

yatch replied:
Thanks! I'll fix it!

sumanpanchal commented:
Simulation error with three node network.
![screenshot from 2017-05-19 15-40-21](https://cloud.githubusercontent.com/assets/12022110/26243569/f5aaa87e-3ca9-11e7-8c66-560f88cf1938.png)

yatch commented:
@sumanpanchal Thank you for your testing! That example, `rpl-tsch-sixtop`, supports only a two-node network. It's not been tested with other setting such as having three nodes.

It seems invalid memory access occurred in your test; "Assertion failed" messages in the log could help, by the way.

awolokita commented:
Hi @yatch,thanks for all your work into 6top. This req has been quiet for a little while now. Are you still working on this? It would be fantastic to have this functionality!
